### PR TITLE
to s3

### LINKED ISSUE #1: dump to s3
as a pipeline user, I want to be able to process data and save them on s3

## Acceptance criteria

* able to save files on s3 

## Tasks

* [x] go throught the okfn/measure codebase and understand how the work 
* [x] write tests 
* [x] write function to put files on s3
* [x] provide README

### Spec

```yaml

my-cool-pipeline:
  title: Cool data
  pipeline:
    -
      run: to_s3
      parameters:
        bucket: mybucket.name
        path: my/s3/path/{owner}/{name}/{version}
```
Note:  it should be possible to pass path exactly in this format as well. The script should be able to get the owner name, package name and the version from datapackage.json and add DataPackage to that path. Eg: executing the pipeline above should put files on `my.bucket.name/my/s3/path/core/finance-vix/lates/...`

* Version defaults to _latest_
* If path includes {owner}, it should present in datapackage.json as well Eg: 
```
{"name": "finance-vix", "owner": "core"}
```

### PR DESCRIPTION
**Pull request by zelima**
This pull request fixes #1 

* [x] I've added tests to cover the proposed changes
  * partially - see note

Changes proposed in this pull request:

- helper function to format path (Key) to the file on s3
- a processor for putting files on s3
- deps updated for test

@akariv Note:

The PR is not complete - the tests are not checking if resources are on s3 as well, cause I could not find an easy way to pass the resource iterator to `mock_processor_test`. Some help would be useful

### CODE REVIEW

File: datapackage_pipelines_aws/helpers.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import os
+import logging
+
+def generate_path(file_path, base_path='', datapackage={}):
+    vers = datapackage.get('version', 'latest')
```
Comment (akariv) [inline datapackage_pipelines_aws/helpers.py]: This feels a bit hard-coded to me.

We could use a simpler (and more powerful):
`path = base_path.format(**self.datapackage)`
no?

File: datapackage_pipelines_aws/processors/to_s3.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import os
+
+from datapackage_pipelines.lib.dump.dumper_base import CSVDumper
+from datapackage_pipelines_aws import helpers
+
+class S3Dumper(CSVDumper):
+
+    def initialize(self, params):
+        super(S3Dumper, self).initialize(params)
+        self.bucket = params['bucket']
```
Comment (akariv) [inline datapackage_pipelines_aws/processors/to_s3.py:11]: What about AWS access key and secret?

File: datapackage_pipelines_aws/helpers.py (hunk 1 above)
Comment (akariv) [reply datapackage_pipelines_aws/helpers.py]: or even:
```python
format_params = {'version': 'latest'}
format_params.update(self.datapackage)
base_path.format(**format_params)
```

File: tests/test_to_s3.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    @mock_s3
+    def test_puts_datapackage_on_s3(self):
+        s3 = boto3.resource('s3')
+        s3.create_bucket(Bucket=self.bucket)
+
+        mock_processor_test(self.processor_path,
+                            (self.params,
+                            self.datapackage,
+                            iter([])))
```
Comment (akariv) [inline tests/test_to_s3.py]: This should be a list of lists of the resource data. e.g.:

```
[ 
  [ {'Name': 'a', 'Date': datetime.datetime.now().date()}, ... 
  ] 
]
```

**Review by akariv — CHANGES_REQUESTED** [review]

File: datapackage_pipelines_aws/processors/to_s3.py (hunk 2 above)
Comment (zelima) [reply datapackage_pipelines_aws/processors/to_s3.py:11]: @akariv boto3 takes care of credential if they are set-up. It looks up in aws config file or searches for ENV variables http://boto3.readthedocs.io/en/latest/guide/configuration.html#aws-config-file
But we can have them as a part of the spec as well if that is a case

File: tests/test_to_s3.py (hunk 3 above)
Comment (zelima) [reply tests/test_to_s3.py]: I'm getting this error in that case
```
def handle_resources(self, datapackage,
                         resource_iterator,
                         parameters, stats):
        datapackage['count_of_rows'] = 0
        datapackage['hash'] = hashlib.md5()
        for resource in resource_iterator:
>           resource_spec = resource.spec
E           AttributeError: 'list' object has no attribute 'spec'
```

Comment (akariv) [reply tests/test_to_s3.py]: Okay, try this:
```
res =  [ {'Name': 'a', 'Date': datetime.datetime.now().date()}, ... ] 
res.spec = datapackage['resources'][0]
res_iter = [res]
```

We should probably find a wrapper for that.

File: tox.ini [hunk 4]
```diff
@@ -6,8 +6,8 @@ envlist=
 
 [testenv]
 deps=
-  mock
-  requests-mock
+	google-compute-engine
```
Comment (akariv) [inline tox.ini]: google-compute-engine?

Comment (zelima) [reply tox.ini]: boto was complaining on Travis 
```
Traceback:
tests/test_to_s3.py:5: in <module>
    from moto import mock_s3
.tox/py/lib/python3.6/site-packages/moto/__init__.py:9: in <module>
    from .autoscaling import mock_autoscaling, mock_autoscaling_deprecated  # flake8: noqa
.tox/py/lib/python3.6/site-packages/moto/autoscaling/__init__.py:2: in <module>
    from .models import autoscaling_backends
.tox/py/lib/python3.6/site-packages/moto/autoscaling/models.py:2: in <module>
    from boto.ec2.blockdevicemapping import BlockDeviceType, BlockDeviceMapping
.tox/py/lib/python3.6/site-packages/boto/__init__.py:1216: in <module>
    boto.plugin.load_plugins(config)
.tox/py/lib/python3.6/site-packages/boto/plugin.py:93: in load_plugins
    _import_module(file)
.tox/py/lib/python3.6/site-packages/boto/plugin.py:75: in _import_module
    return imp.load_module(name, file, filename, data)
.tox/py/lib/python3.6/imp.py:234: in load_module
    return load_source(name, filename, file)
.tox/py/lib/python3.6/imp.py:172: in load_source
    module = _load(spec)
/usr/lib/python2.7/dist-packages/google_compute_engine/boto/compute_auth.py:19: in <module>
    from google_compute_engine import logger
E   ModuleNotFoundError: No module named 'google_compute_engine'
```
This answer summarises the reason https://github.com/GoogleCloudPlatform/compute-image-packages/issues/262#issuecomment-225736860

**Review by akariv — APPROVED** [review]

**Pull request merged**